core/src/main/scala/akka/kafka/internal/BaseSingleSourceLogic.scala

```diff
@@ -79,18 +84,24 @@ import scala.concurrent.{ExecutionContext, Future}
       tps ++= topics.keySet
   }
 
-  protected def filterRevokedPartitions(topicPartitions: Set[TopicPartition]): Unit =
-    if (topicPartitions.nonEmpty) {
-      log.debug("filtering out messages from revoked partitions {}", topicPartitions)
-      // as buffer is an Iterator the filtering will be applied during `pump`
-      buffer = buffer.filterNot { record =>
-        val tp = new TopicPartition(record.topic, record.partition)
-        topicPartitions.contains(tp)
-      }
-    }
+  protected def partitionAssignedHandler(assignedTps: Set[TopicPartition]): Unit = {
+    tps ++= assignedTps
+    log.debug("Assigned partitions: {}. All partitions: {}", assignedTps, tps)
+    requestMessages()
+  }
+
+  protected def partitionRevokedHandler(revokedTps: Set[TopicPartition]): Unit = {
+    tps --= revokedTps
+    log.debug("Revoked partitions: {}. All partitions: {}", revokedTps, tps)
+  }
+
+  protected def partitionLostHandler(lostTps: Set[TopicPartition]): Unit = {
+    tps --= lostTps
+    log.debug("Lost partitions: {}. All partitions: {}", lostTps, tps)
+  }
```

Comment:
These can be private or even collapsed into the callback creation.

Reply:
Good idea.